[DPE-3253] Charmed Spark tutorial (#87)
canonical/spark-client-snap: docs/tutorial/t-spark-monitoring.md

> ## Monitoring the Spark Cluster

Reviewer: ```suggestion
## Monitoring the Spark cluster
```

> By default, Spark stores the logs of drivers and executors as pod logs in the local file system, which gets lost if the pods are deleted. Spark provides us with the feature to store these logs in S3 so that they can later be retrieved and visualized.

Reviewer 2: ```suggestion
By default, Spark stores the logs of drivers and executors as pod logs in the local file system, which are lost oncepods are deleted. Spark provides us with the feature to store these logs in a persistent object storage, like S3, so that they can later be retrieved and visualized by a componed named Spark History server.
```

Reviewer 2: Uhm, the text above only applies to SHS. I would probably move it to the Spark History Server subsection

> Monitoring of Spark cluster can be done in two ways,

Reviewer 2: ```suggestion
Charmed Spark provides native support and integration for two monitoring systems:
```

1. Using Spark History Server 

> 2. Using Canonical Observability Stack (COS)

Reviewer 2: I believe it makes sense to explain to the user that these two tools address two different kind of users, see [here](https://github.com/deusebio/gitex-2023-charmed-spark?tab=readme-ov-file#fetch-your-endpoints)

Author: Done.

> Let's explore each one of these.

Reviewer 3: ```suggestion
Let's explore each one of these options.
```

### Monitoring with Spark History Server

Spark History Server is a user interface to monitor the metrics and performance of completed and running Spark applications. The Spark History Server is offered as a charm in the Charmed Spark solution, which can be deployed via Juju.

> 

Reviewer: I think it would be nice to have a small breakdown of the steps to expect, similar to what you did in the next section about COS. 
E.g.
1. Set up the Spark History Server with Juju
2. Connect SHS to the S3 bucket
3. Access the web UI to view and monitor jobs

Author: This breakdown is provided in the paragraph that follows just after the `juju add-model` command.

```bash
# Get config from old service account and store in a file
spark-client.service-account-registry get-config \
  --username spark --namespace spark > properties.conf

# Add a few more config options for the storage of event logs
echo "spark.eventLog.enabled=true" >> properties.conf
echo "spark.eventLog.dir=s3a://spark-tutorial/spark-events/" >> properties.conf
echo "spark.history.fs.logDirectory=s3a://spark-tutorial/spark-events/" >> properties.conf

# Create a new service account and load configurations from the file
spark-client.service-account-registry create \
  --username spark --namespace history-server \
  --properties-file properties.conf
```

We've configured Spark to write event logs to `spark-events` directory in the `spark-tutorial` bucket. However, that directory doesn't exist yet. Let's create that directory in S3.

```bash
aws s3api put-object --bucket spark-tutorial --key spark-events/ # / at the end is requied

> ```

Reviewer: Inline comments are generally discouraged in most style guides. I'd write that detail as a note before or after the command, e.g.

"Let's create that directory in S3. (Note that the `/` at the end is required)"

Author: Fixed.

Now, let's deploy the [`spark-history-server-k8s`](https://github.com/canonical/spark-history-server-k8s-operator) charm in our Juju model. 

```bash

> juju deploy spark-history-server-k8s -n1 --channel 3.4/stable

Reviewer 3: Probably I will deploy from `edge
`

Reviewer 2: Actually, ideally, we should produce documentation for stable branches if possible, as edge channels may be unstable and I don't want people testing out things to hit road blocks

Author: Keeping up with `3.4/stable` for now.

s3-integrator                      active      1  s3-integrator             edge         14  10.152.183.87   no       
spark-history-server-k8s           active      1  spark-history-server-k8s  3.4/stable   15  10.152.183.111  no       

Unit                         Workload  Agent  Address      Ports  Message
s3-integrator/0*             active    idle   10.1.29.167         
spark-history-server-k8s/0*  active    idle   10.1.29.169         

Integration provider               Requirer                                 Interface            Type     Message
s3-integrator:s3-credentials       spark-history-server-k8s:s3-credentials  s3                   regular  
s3-integrator:s3-integrator-peers  s3-integrator:s3-integrator-peers        s3-integrator-peers  peer   
```

The Spark History Server is now successfully configured to read event logs from S3 bucket. Let's run a simple job so that Spark can generate some event logs. We're going to use the same `count_vowels.py` example we used earlier, which already exists in the `spark-tutorial` bucket.

```bash
spark-client.spark-submit \

>     --username spark --namespace history-server \

Reviewer 3: It is not necessary to run the job in the same namespace

Author: During my testing, using different namespaces resulted in the DNS of the pods (running in Juju model's namespace) not being resolved from the pods running in spark serviceaccount's namespace. That's the reason I kept dedicated namespaces for each section of the tutorial.

    --deploy-mode cluster \
    s3a://spark-tutorial/count_vowels.py
```

The Spark History Server comes with a Web UI for us to view and monitor the jobs submitted to Spark Cluster. The web UI can be accessed at port 18080 of the IP address of the `spark-history-server-k8s/0` unit. However, it is a good practice to access it via Ingress rather than directly accessing the unit IP address. Using an Ingress will allow us to have a common entrypoint to the apps running in the Juju model. We can add an Ingress by deploying and integrating [`traefik-k8s`](https://charmhub.io/traefik-k8s) charm with `spark-history-server-k8s`.

```bash
# Deploy traefik charm

> juju deploy traefik-k8s --channel latest/candidate --trust

Reviewer 3: also here `edge`

Author: Using channel `latest/stable` as per the earlier comment.

loki:metrics-endpoint               prometheus:metrics-endpoint  prometheus_scrape      regular  
loki:replicas                       loki:replicas                loki_replica           peer     
prometheus:grafana-dashboard        grafana:grafana-dashboard    grafana_dashboard      regular  
prometheus:grafana-source           grafana:grafana-source       grafana_datasource     regular  
prometheus:prometheus-peers         prometheus:prometheus-peers  prometheus_peers       peer     
traefik:ingress                     alertmanager:ingress         ingress                regular  
traefik:ingress                     catalogue:ingress            ingress                regular  
traefik:ingress-per-unit            loki:ingress                 ingress_per_unit       regular  
traefik:ingress-per-unit            prometheus:ingress           ingress_per_unit       regular  
traefik:metrics-endpoint            prometheus:metrics-endpoint  prometheus_scrape      regular  
traefik:peers                       traefik:peers                traefik_peers          peer     
traefik:traefik-route               grafana:ingress              traefik_route          regular  
```

> At this point, the observability stack has been deployed but Charmed Spark knows nothing about it. Since Charmed Spark comes with a built in spark metrics exporter that exports metric logs to a Prometheus gateway, we need to deploy a Prometheus gateway and then add Spark options to connect to it. The Prometheus gateway in turn will then be integrated with Prometheus. Let's deploy `prometheus-pushgateway-k8s` charm and integrate it with `prometheus` charm.

Reviewer 2: ```suggestion
At this point, the observability stack has been deployed but Charmed Spark is not wired to it yet. Generally, Prometheus collects metrics of services by regularly scraping dedicated endpoints. However, Spark jobs can be ephemeral processes that may last not so long to be scraped. Moreover, the IPs/hostnames of pods are likely to change among runs, therefore requiring a self-discoverable automation. This is the reason why we decided to provide support for jobs run with Charmed Spark to push metrics (rather than having Prometheus pulling them) to a caching service, called Prometheus pushgateway, that stores the metrics and exposed them to Prometheus for regular scraping, even when the Spark job has finished. Therefore, to wire Spark jobs with the observability stack, we need to deploy a Prometheus pushgateway and then add Spark options to connect to it. The Prometheus pushgateway in turn will then be integrated with Prometheus. Let's deploy `prometheus-pushgateway-k8s` charm and integrate it with `prometheus` charm.
```

```bash
juju deploy prometheus-pushgateway-k8s --channel edge

juju integrate prometheus-pushgateway-k8s prometheus
```

Now, for Spark to be able to access the Prometheus gateway, we need the gateway address and port. Let's export them as environment variables so that they can be used later.

```shell
# Get prometheus gateway IP
export PROMETHEUS_GATEWAY=$(juju status --format=json | jq -r '.applications."prometheus-pushgateway-k8s".address') 

export PROMETHEUS_PORT=9091
```

Now that we have Prometheus gateway IP adress and port, let's create a new service account in `cos` namespace with all configuration options as `spark` service account in `spark` namespace with a few additional configuration options related to Prometheus gateway.

```shell
# Get config from old service account and store in a file
spark-client.service-account-registry get-config \
  --username spark --namespace spark > properties.conf

# Create a new service account and load configurations from the file
spark-client.service-account-registry create \
  --username spark --namespace cos \

>   --properties-file properties.conf

Reviewer 2: Also in this case, it should not be need to deploy spark-jobs and service account to the cos namespace. Actually, maybe it is better to have a namespace for observability stuff, and another way for the user to run spark jobs

Author: During my testing, using different namespaces resulted in the DNS of the pods (running in Juju model's namespace) not being resolved from the pods running in spark serviceaccount's namespace. That's the reason I kept dedicated namespaces for each section of the tutorial.
